feature_engine/timeseries/forecasting/window_features.py

```diff
@@ -0,0 +1,333 @@
+        self.variables_ = _find_or_check_numerical_variables(X, self.variables)
+
+        # check if dataset contains na
+        if self.missing_values == "raise":
+            _check_contains_na(X, self.variables_)
+            _check_contains_inf(X, self.variables_)
+
+        self.feature_names_in_ = X.columns.tolist()
+        self.n_features_in_ = X.shape[1]
+
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Adds window features.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The data to transform.
+
+        Returns
+        -------
+        X_new: Pandas dataframe, shape = [n_samples, n_features + lag_features]
+            The dataframe with the original plus the new variables.
+        """
+        # checmk method fit has been called
+        check_is_fitted(self)
+
+        # check if 'X' is a dataframe
+        X = _is_dataframe(X)
+
+        # check if input data contains the same number of columns as the fitted dataframe.
+        _check_input_matches_training_df(X, self.n_features_in_)
+
+        # Dataframes must have unique values in the index and no missing data.
+        # Otherwise, when we merge the lag features we will duplicate rows.
+        if X.index.isnull().sum() > 0:
+            raise NotImplementedError(
+                "The dataframe's index contains NaN values or missing data. "
+                "Only dataframes with complete indexes are compatible with "
+                "this transformer."
+            )
+
+        if X.index.is_unique is False:
+            raise NotImplementedError(
+                "The dataframe's index does not contain unique values. "
+                "Only dataframes with unique values in the index are compatible "
+                "with this transformer."
+            )
+
+        # check if dataset contains na
+        if self.missing_values == "raise":
+            _check_contains_na(X, self.variables_)
+            _check_contains_inf(X, self.variables_)
+
+        if self.sort_index is True:
+            X.sort_index(inplace=True)
+
+        # if freq is not None, it overrides periods.
+        if self.freq is not None:
+            tmp = (X[self.variables_]
+                   .rolling(window=self.window).apply(self.function)
+                   .shift(freq=self.freq)
+                   )
+
+        else:
+            tmp = (X[self.variables_]
+                   .rolling(window=self.window).apply(self.function)
+                   .shift(periods=self.periods)
+                   )
+
+        tmp.columns = self.get_feature_names_out(self.variables_)
+
+        X = X.merge(tmp, left_index=True, right_index=True, how="left")
+
+        if self.drop_original:
+            X = X.drop(self.variables_, axis=1)
+
+        return X
+
+    def get_feature_names_out(self, input_features: Optional[List] = None) -> List:
```

Comment:
Is this method identical to lag features? if yes, then let's move it to the base class.

Reply:
It's different. It includes the window in the name, e.g. var_A_window_2_freq_3d and var_B_window_3_periods_2.

I guess we could move the following code to the BaseForecast class:

```
check_is_fitted(self)

        # create names for all window features or just the indicated ones.
        if input_features is None:
            input_features_ = self.variables_
        else:
            if not isinstance(input_features, list):
                raise ValueError(
                    f"input_features must be a list. Got {input_features} instead."
                )
            if any([f for f in input_features if f not in self.variables_]):
                raise ValueError(
                    "Some features in input_features were not transformed. This method only "
                    "provides the names of the transform features with this method."
                )
            # create just indicated window features
            input_features_ = input_features
```

Reply:
I modified this method in https://github.com/Morgan-Sell/feature_engine/pull/9